packages/typespec-client-generator-core/design-docs/multiple-services.md

```diff
@@ -0,0 +1,176 @@
+# Multiple Service Support in TypeSpec Client Generator Core
+
+## Background
+
+Previously, TCGC [client](./client.md) only supported generating a client from a single service. However, in real-world scenarios, a single package often contains multiple services. TCGC must support multiple services within one package to address these needs.
+
+## User Scenario
+
+1. Merging multiple services' namespaces into one client
+
+This scenario is common in Azure management services. For example, the compute team maintains several services: `Compute`, `Disk`, `Gallery`, and `Sku`. These services share the same endpoint and credential but have different versioning. When migrating these services into TypeSpec, services team wants to follow the existing way to generate SDK: geneate one SDK with a single client that could manage all these services with different versioning, instead of generate multiple SDKs for these multiple services. Therefore, TCGC must support auto-merging operations and nested namespaces/interfaces from multiple services with different versioning into one client.
+
+For example, given two services `Disk` and `Gallery`, with Python SDK, the generated client code should look like:
+
+```python
+client = ComputeManagementClient(credential=DefaultAzureCredential(), subscription_id="{subscription-id}")
+client.disks.list_by_resource_group(resource_group_name="myResourceGroup") # this operation will use API version defined in Disk service
+client.galleries.list(location="eastus") # this operation will use API version defined in Gallery service
+```
+
+## First Step Design
+
+The initial design focuses on the scenario of auto-merging multiple services into a single client. For other scenarios, such as redefining client hierarchy for multiple services, we will consider them in future iterations.
+
+### Syntax Proposal
+
+Previously, the `service` property of the `@client` option only supported a single service. We propose extending it to accept an array of services.
+
+For example, given two services, the original service specs are:
+
+```typespec title="ServiceA/main.tsp"
+@service
+@versioned(VersionsA)
+namespace ServiceA;
+
+enum VersionsA {
+  av1,
+  av2,
+}
+interface AI {
+  @route("/aTest")
+  aTest(@query("api-version") apiVersion: VersionsA): void;
+}
+```
+
+```typespec title="ServiceB/main.tsp"
+@service
+@versioned(VersionsB)
+namespace ServiceB;
+
+enum VersionsB {
+  bv1,
+  bv2,
+}
+interface BI {
+  @route("/bTest")
+  bTest(@query("api-version") apiVersion: VersionsB): void;
+}
+```
+
+To define a combined client:
+
+```typespec title="client.tsp"
+import "./ServiceA/main.tsp";
+import "./ServiceB/main.tsp";
+import "@azure-tools/typespec-client-generator-core";
+
+using Azure.ClientGenerator.Core;
+
+@client({
+  name: "CombineClient",
+  service: [ServiceA, ServiceB],
+})
+@useDependency(ServiceA.VersionsA.av2, ServiceB.VersionsB.bv2) // optional
+namespace CombineClient;
```

Comment:
Trying to map this case to compute one, do we have followings?
- CombineClient => Compute;
- ServiceA => Compute.Disk;
- ServiceB => Compute.Gallary;

So we need to update the existing typespec?

Reply:
Yes. Original one.